### PR TITLE
[SDK][APPLICATIONS][DRIVERS] Unlink ntdll.dll and use modified <debug.h>

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
We decided that the excessive `ntdll.dll` dependency of the user-mode programs should be reduced if possible.

JIRA issue: [CORE-17969](https://jira.reactos.org/browse/CORE-17969)
Revival of PR #4256

## Proposed changes

- Modify `<debug.h>` to support user-mode debugging.
- Define `NTOS_MODE_USER` macro in the top of some modules.
- Unlink `ntdll.dll` of some modules.

## Screenshots
Using `gdb`:
after

### CODE REVIEW

File: base/applications/mspaint/precomp.h [hunk 1]
```diff
@@ -21,8 +21,9 @@
 #include <shellapi.h>
 #include <htmlhelp.h>
 
+#define NO_STRSAFE
```
Comment (HBelusca) [inline base/applications/mspaint/precomp.h:24]: What's the problem with using stringsafe here?

File: sdk/include/reactos/debug/rappdbg.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+static inline ULONG rappdbg_printfA(const char *file, int line, const char *fmt, ...)
+{
+    char buf[APPDBG_BUFSIZE];
+    va_list va;
+    const char *fname;
+    size_t cch;
+
+    va_start(va, fmt);
+
+    fname = strrchr(file, '\\');
```
Comment (HBelusca) [inline sdk/include/reactos/debug/rappdbg.h:44]: Why doing these conversions?

Comment (HBelusca) [inline sdk/include/reactos/debug/rappdbg.h:44]: Why not keeping the original file path as it comes out? (this is what the regular DPRINT etc. macros do)

File: base/applications/mspaint/precomp.h (hunk 1 above)
Comment (katahiromz) [reply base/applications/mspaint/precomp.h:25]: The `<strsafe.h>` implementation can refuse incompatible function calls.

https://github.com/reactos/reactos/blob/47f6745bcdfb833b0311e052f04e6cd5634660eb/sdk/include/psdk/strsafe.h#L1921

So, some programs are incompatible to `<strsafe.h>`.

File: sdk/include/reactos/debug/rappdbg.h (hunk 2 above)
Comment (katahiromz) [reply sdk/include/reactos/debug/rappdbg.h]: Fixed.

File: sdk/include/reactos/debug.h [hunk 3]
```diff
@@ -58,6 +60,44 @@ RtlAssert(
     _In_opt_z_ PCHAR Message
 );
 
+#else /* def NTOS_MODE_USER */
+
+typedef _Return_type_success_(return >= 0) LONG NTSTATUS;
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:65]: For what code is the NTSTATUS needed?

File: sdk/include/reactos/debug.h [hunk 4]
```diff
@@ -30,6 +30,8 @@
 extern "C" {
 #endif
 
+#ifndef NTOS_MODE_USER
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:33]: Actually it would be nice to keep using the original DbgPrint and RtlAssert in case `NTOS_MODE_USER` is defined: this means the (user-mode) app will use ntdll functions too, and so it can call these direct ntdll-exported debug functions.

I would suggest something like this (adapted from a similar situation for the `TerminateCurrentProcess` macro at the end of this file):
```c
#if defined(_NTDDK_) || defined(NTOS_MODE_USER)
/* Kernel-Mode, or Native User-Mode */

// Here, the existing DbgPrint etc. definitions...

#elif defined(_WIN32)
/* Win32 User-Mode */

// Here, your new function and definitions...

#else
    #error "Unsupported target."
#endif
```

File: sdk/include/reactos/debug.h [hunk 5]
```diff
@@ -58,6 +60,78 @@ RtlAssert(
...
+          "File '%s', Line %ld:\n\n%s\n\n%s",
+          (LPCSTR)FileName, LineNumber, (LPCSTR)FailedAssertion, Message);
+#endif
+
+    id = MessageBoxA(NULL, szBuff, "Assertion Failure", MB_ICONERROR | MB_ABORTRETRYIGNORE);
+
+    if (id == IDABORT)
+        ExitProcess((UINT)-1);
+
+    if (id == IDRETRY)
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:123]:
```suggestion
    else if (id == IDRETRY)
```

File: sdk/include/reactos/debug.h [hunk 6]
```diff
@@ -58,6 +60,78 @@ RtlAssert(
...
+__cdecl
+W32DbgPrint(_In_z_ _Printf_format_string_ PCSTR Format, ...)
+{
+    va_list va;
+    char szBuff[512];
+
+    va_start(va, Format);
+
+#ifdef NO_STRSAFE
+    wvsprintfA(szBuff, Format, va);
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:82]: Why not using the standard C runtime function _vsnprintf instead? (This wvsprintfA would enforce a dependence on user32.dll, which is not always present for win32 user-mode apps.)

File: sdk/include/reactos/debug.h (hunk 3 above)
Comment (katahiromz) [reply sdk/include/reactos/debug.h]: Deleted.

File: sdk/include/reactos/debug.h (hunk 6 above)
Comment (katahiromz) [reply sdk/include/reactos/debug.h]: Used `_vsnprintf`.

File: sdk/include/reactos/debug.h (hunk 4 above)
Comment (katahiromz) [reply sdk/include/reactos/debug.h]: OK.

File: sdk/include/reactos/debug.h [hunk 7]
```diff
@@ -18,6 +18,10 @@
 #define __RELFILE__ __FILE__
 #endif
 
+#ifdef NDEBUG
+#undef DBG
+#endif
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:23]: Why doing this? `DBG` is the general macro that tells whether a program is compiled in debug mode. `˙NDEBUG` means, that we don't want the debug macros to show debug output when using debug.h.

Comment (katahiromz) [reply sdk/include/reactos/debug.h:23]: Some codes manually define `NDEBUG` and disable the debugging output... I followed this manner.

https://github.com/reactos/reactos/blob/47f6745bcdfb833b0311e052f04e6cd5634660eb/win32ss/gdi/ntgdi/freetype.c#L36-L37

https://github.com/reactos/reactos/blob/47f6745bcdfb833b0311e052f04e6cd5634660eb/win32ss/gdi/ntgdi/bitblt.c#L10-L11

Comment (HBelusca) [reply sdk/include/reactos/debug.h:23]: Yes but this means, that if they do an explicit `DPRINT1` (not `DPRINT`) then the DPRINT1 output will be output, which is what is expected. When `NDEBUG` is not defined, which is e.g. usually done when testing something locally, both `DPRINT` and `DPRINT1` will be output (this is one thing I forgot to tell in the previous message). But in all cases, the generic `DBG` define symbol still remains as it is (untouched).

File: base/applications/mspaint/dialogs.cpp [hunk 8]
```diff
@@ -320,15 +320,15 @@ void CFontsDialog::InitFontSizes()
...
+        StringCchPrintf(szText, _countof(szText), TEXT("%d"), s_sizes[i]);
         INT iItem = ComboBox_AddString(hwndSizes, szText);
         if (s_sizes[i] == (INT)registrySettings.PointSize)
             ComboBox_SetCurSel(hwndSizes, iItem);
     }
 
     if (ComboBox_GetCurSel(hwndSizes) == CB_ERR)
     {
-        wsprintf(szText, TEXT("%d"), (INT)registrySettings.PointSize);
+        StringCchPrintf(szText, _countof(szText), TEXT("%d"), (INT)registrySettings.PointSize);
```
Comment (HBelusca) [inline base/applications/mspaint/dialogs.cpp:331]: I would suggest you do that in a very separate commit, outside of this PR.

Comment (katahiromz) [reply base/applications/mspaint/dialogs.cpp:331]: Without these changes, the build will fail.

Comment (HBelusca) [reply base/applications/mspaint/dialogs.cpp:331]: ah yes, the mspaint would need an explicit `#include <strsafe.h>` in its precompiled header.

File: sdk/include/reactos/debug.h [hunk 9]
```diff
@@ -58,6 +61,73 @@ RtlAssert(
...
+#include <strsafe.h>
+#endif
+
+static inline
+ULONG
+__cdecl
+W32DbgPrint(_In_z_ _Printf_format_string_ PCSTR Format, ...)
+{
+    va_list va;
+    char szBuff[512];
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:77]: The idea of a user-adjustable `APPDBG_BUFSIZE` value (used here, and in the W32Assert below) sounded nice.

Comment (katahiromz) [reply sdk/include/reactos/debug.h]: `DEBUG_BUFSIZE`

File: sdk/include/reactos/debug.h [hunk 10]
```diff
@@ -58,6 +61,77 @@ RtlAssert(
...
+{
+    INT id;
+    char szBuff[DEBUG_BUFSIZE];
+
+#ifdef NO_STRSAFE
+    _snprintf(szBuff, _countof(szBuff),
+#else
+    StringCchPrintfA(szBuff, _countof(szBuff),
+#endif
+                     "File '%s',\nLine %ld:\n\n%s\n\n%s",
```
Comment (katahiromz) [inline sdk/include/reactos/debug.h:118]: Is this assertion message alright?

Comment (HBelusca) [reply sdk/include/reactos/debug.h]: For information, here is what the ntdll RtlAssert calls when displaying the assertion:
```c
    /* Print the assertion */
    DbgPrint("\n*** Assertion failed: %s%s\n"
             "***   Source File: %s, line %lu\n\n",
             Message != NULL ? Message : "",
             (PSTR)FailedAssertion,
             (PSTR)FileName,
             LineNumber);
```
(note the `Message` NULL check)

Comment (HBelusca) [reply sdk/include/reactos/debug.h]: You can also observe that the win32 CRT has equivalent functions, and you may be inspired by its assertion message:
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/crt/misc/dbgrpt.cpp;hb=47f6745bcdfb833b0311e052f04e6cd5634660eb#l79
(Note: these can only be used in a win32 user-mode app that explicitly uses the CRT.)

File: sdk/include/reactos/debug.h [hunk 11]
```diff
@@ -58,6 +61,73 @@ RtlAssert(
...
+W32Assert(
+    _In_ PVOID FailedAssertion,
+    _In_ PVOID FileName,
+    _In_ ULONG LineNumber,
+    _In_opt_z_ PCHAR Message)
+{
+    INT id;
+    char szBuff[DEBUG_BUFSIZE];
+
+#ifdef _STRSAFE_H_INCLUDED_
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:109]: Nice :)

File: sdk/include/reactos/debug.h [hunk 12]
```diff
@@ -58,6 +61,73 @@ RtlAssert(
...
+#else
+    _snprintf(szBuff, _countof(szBuff),
+#endif
+              "File '%s',\nLine %ld:\n\n%s\n\n%s",
+              (LPCSTR)FileName, LineNumber, (LPCSTR)FailedAssertion, Message);
+
+    id = MessageBoxA(NULL, szBuff, "Assertion Failure", MB_ICONERROR | MB_ABORTRETRYIGNORE);
+
+    if (id == IDABORT)
+        ExitProcess((UINT)-1);
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:120]: or use `ERROR_ASSERTION_FAILURE`

File: sdk/include/reactos/debug.h [hunk 13]
```diff
@@ -58,6 +61,73 @@ RtlAssert(
...
+    INT id;
+    char szBuff[DEBUG_BUFSIZE];
+
+#ifdef _STRSAFE_H_INCLUDED_
+    StringCchPrintfA(szBuff, _countof(szBuff),
+#else
+    _snprintf(szBuff, _countof(szBuff),
+#endif
+              "File '%s',\nLine %ld:\n\n%s\n\n%s",
+              (LPCSTR)FileName, LineNumber, (LPCSTR)FailedAssertion, Message);
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:115]:
```suggestion
              (PCSTR)FileName, LineNumber, (PCSTR)FailedAssertion, Message);
```

File: base/applications/mspaint/dialogs.cpp (hunk 8 above)
Comment (HBelusca) [reply base/applications/mspaint/dialogs.cpp:331]: OK now, you can move both this change and the one in the mspaint's precompiler header into a separate commit, I guess... ? 😄

File: sdk/include/reactos/debug.h (hunk 13 above)
Comment (katahiromz) [reply sdk/include/reactos/debug.h]: OK.

File: sdk/include/reactos/debug.h (hunk 12 above)
Comment (katahiromz) [reply sdk/include/reactos/debug.h]: Used.

File: sdk/include/reactos/debug.h [hunk 14]
```diff
@@ -58,6 +61,74 @@ RtlAssert(
...
+#ifdef _STRSAFE_H_INCLUDED_
+    StringCchPrintfA(szBuff, _countof(szBuff),
+#else
+    _snprintf(szBuff, _countof(szBuff),
+#endif
+              "*** Assertion failed: %s%s\n"
+              "***   Source File: %s, line %lu",
+              (Message ? Message : ""), (PSTR)FailedAssertion, (PSTR)FileName, LineNumber);
+
+    id = MessageBoxA(NULL, szBuff, "Assertion Failure", MB_ICONERROR | MB_ABORTRETRYIGNORE);
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:118]: For the messagebox flags I would suggest: `MB_ABORTRETRYIGNORE | MB_ICONERROR | MB_SETFOREGROUND | MB_TASKMODAL` (like in the CRT debugging).
The 3rd flag puts the msgbox in front when it appears, and the last flag makes that "all the top-level windows belonging to the current thread [of the current application] are disabled if the hWnd parameter is NULL", and since the app is crashing, it should "prevent input to other windows in the calling thread without suspending other threads."
(I'm quoting MSDN here.)

File: sdk/include/reactos/debug.h [hunk 15]
```diff
@@ -58,6 +61,74 @@ RtlAssert(
...
+#ifdef _STRSAFE_H_INCLUDED_
+    StringCchVPrintfA(szBuff, _countof(szBuff), Format, va);
+#else
+    _vsnprintf(szBuff, _countof(szBuff), Format, va);
+#endif
+
+    OutputDebugStringA(szBuff);
+    va_end(va);
+
+    return 0; /* Return STATUS_SUCCESS, since we are supposed to mimic DbgPrint */
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:90]: Win32 equivalent is `ERROR_SUCCESS`.

File: sdk/include/reactos/debug.h [hunk 16]
```diff
@@ -58,6 +61,74 @@ RtlAssert(
...
+
+#ifdef _STRSAFE_H_INCLUDED_
+    StringCchPrintfA(szBuff, _countof(szBuff),
+#else
+    _snprintf(szBuff, _countof(szBuff),
+#endif
+              "*** Assertion failed: %s%s\n"
+              "***   Source File: %s, line %lu",
+              (Message ? Message : ""), (PCSTR)FailedAssertion, (PCSTR)FileName, LineNumber);
+
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:117]: It could be of interest to also send this message to dprint (besides popping up the message box) in case the application runs under a debugger.

File: sdk/include/reactos/debug.h [hunk 17]
```diff
@@ -58,6 +61,78 @@ RtlAssert(
...
+    StringCchPrintfA(szBuff, _countof(szBuff),
+#else
+    _snprintf(szBuff, _countof(szBuff),
+#endif
+              "*** Assertion failed: %s%s\n"
+              "***   Source File: %s, line %lu\n\n"
+              "(Press Retry to debug the application)\n",
+              (Message ? Message : ""), (PCSTR)FailedAssertion, (PCSTR)FileName, LineNumber);
+
+    OutputDebugStringA(szBuff);
```
Comment (HBelusca) [inline sdk/include/reactos/debug.h:119]: This is what I was thinking about regarding "sending this message to dprint in case the app runs under a debugger":
```suggestion
    if (IsDebuggerPresent())
        OutputDebugStringA(szBuff);
```
(inspired from the CRT debug helpers)

File: base/applications/mspaint/dialogs.cpp (hunk 8 above)
Comment (katahiromz) [reply base/applications/mspaint/dialogs.cpp]: Reverted.

File: dll/cpl/mmsys/mmsys.h [hunk 18]
```diff
@@ -6,6 +6,7 @@
 #define WIN32_NO_STATUS
 #define _INC_WINDOWS
 #define COM_NO_WINDOWS_H
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline dll/cpl/mmsys/mmsys.h:9]: The CPL is a GUI program, it should not use the NTDLL debugging stuff.

File: win32ss/drivers/miniport/vbe/vbemp.h [hunk 19]
```diff
@@ -23,6 +23,7 @@
 
 /* INCLUDES *******************************************************************/
 
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline win32ss/drivers/miniport/vbe/vbemp.h:26]: No! This is a driver!

File: win32ss/drivers/miniport/pc98vid/pc98vid.h [hunk 20]
```diff
@@ -8,6 +8,7 @@
 #ifndef _PC98VID_PCH_
 #define _PC98VID_PCH_
 
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline win32ss/drivers/miniport/pc98vid/pc98vid.h:11]: No! This is a driver!

File: win32ss/drivers/font/ftfd/ftfd.h [hunk 21]
```diff
@@ -8,6 +8,7 @@
 #ifndef _FTFD_PCH_
 #define _FTFD_PCH_
 
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline win32ss/drivers/font/ftfd/ftfd.h:11]: No! This is a driver!

File: win32ss/drivers/miniport/vga/vgamp.h [hunk 22]
```diff
@@ -23,6 +23,7 @@
 
 /* INCLUDES *******************************************************************/
 
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline win32ss/drivers/miniport/vga/vgamp.h:26]: No! This is a driver!

File: win32ss/drivers/miniport/vga_new/vga.h [hunk 23]
```diff
@@ -10,6 +10,7 @@
 #ifndef _VGA_NEW_PCH_
 #define _VGA_NEW_PCH_
 
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline win32ss/drivers/miniport/vga_new/vga.h:13]: No! This is a driver!

File: win32ss/drivers/miniport/vmx_svga/precomp.h [hunk 24]
```diff
@@ -6,6 +6,7 @@
  * PROGRAMMERS:     ReactOS Portable Systems Group
  */
 
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline win32ss/drivers/miniport/vmx_svga/precomp.h:9]: No! This is a driver!

File: win32ss/drivers/miniport/xboxvmp/xboxvmp.h [hunk 25]
```diff
@@ -15,6 +15,7 @@
  * FIXME: specify headers properly in the triangle brackets and rearrange them
  * in a way so it would be simpler to add NDK and other headers for debugging.
  */
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline win32ss/drivers/miniport/xboxvmp/xboxvmp.h:18]: No! This is a driver!

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: dll/win32/kernel32/k32.h [hunk 26]
```diff
@@ -11,6 +11,7 @@
 
 /* INCLUDES ******************************************************************/
 
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline dll/win32/kernel32/k32.h:14]: Please move this just above the first `#include <ndk/...>` below (line 35)

File: dll/win32/loadperf/loadperf_main.c [hunk 27]
```diff
@@ -18,7 +18,7 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline dll/win32/loadperf/loadperf_main.c:21]: Should not be necessary.

File: sdk/lib/cportlib/cport.c [hunk 28]
```diff
@@ -30,6 +30,7 @@
 
 /* INCLUDES *******************************************************************/
 
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline sdk/lib/cportlib/cport.c:33]: Problematic, as this library is only used in kernel mode.

File: dll/win32/authz/CMakeLists.txt [hunk 29]
```diff
@@ -14,6 +14,7 @@ add_library(authz MODULE
     ${CMAKE_CURRENT_BINARY_DIR}/authz.def)
 
 set_module_type(authz win32dll UNICODE)
-add_importlibs(authz advapi32 msvcrt kernel32 ntdll)
+add_importlibs(authz advapi32 msvcrt kernel32 user32 ntdll)
```
Comment (HBelusca) [inline dll/win32/authz/CMakeLists.txt:17]: This dll should never depend on user32.

File: dll/win32/authz/CMakeLists.txt [hunk 30]
```diff
@@ -14,6 +14,7 @@ add_library(authz MODULE
...
+target_link_libraries(authz wine)
```
Comment (HBelusca) [inline dll/win32/authz/CMakeLists.txt:18]: Should not be needed (we don't use any wine stuff in authz)

File: dll/win32/mswsock/CMakeLists.txt [hunk 31]
```diff
@@ -16,6 +16,6 @@ add_library(mswsock MODULE
     ${CMAKE_CURRENT_BINARY_DIR}/mswsock.def)
 
 set_module_type(mswsock win32dll UNICODE)
-add_importlibs(mswsock ws2_32 advapi32 dnsapi msvcrt kernel32 ntdll)
+add_importlibs(mswsock ws2_32 advapi32 dnsapi msvcrt kernel32 user32)
```
Comment (HBelusca) [inline dll/win32/mswsock/CMakeLists.txt:19]: On Windows: there is a ntdll dependence (it is there because the NT IO functions are used), and user32 is delay-loaded, only when necessary.

File: drivers/wdm/audio/filters/kmixer/kmixer.h [hunk 32]
```diff
@@ -1,6 +1,7 @@
 #ifndef _KMIXER_PCH_
 #define _KMIXER_PCH_
 
+#define NTOS_MODE_USER
```
Comment (HBelusca) [inline drivers/wdm/audio/filters/kmixer/kmixer.h:4]: That's a driver!

**Review by HBelusca — COMMENTED** [review]
Remarks:
- I've added commits that fix aspects of some of the previous commits of this PR.
- the CPORTLIB and kernel32 changes have been manually merged to master: your branch will need to be rebased (and the potential minimal conflicts solved).
- It would be nice to clean up all the "fix build" etc. commits into an intelligible set of commits that touch logically the different modules.

File: sdk/include/reactos/debug.h [hunk 33]
```diff
@@ -18,9 +18,6 @@
 #define __RELFILE__ __FILE__
```
Comment (learn-more) [inline sdk/include/reactos/debug.h:18]: I modified this header to only use ntdll, and not kernel32.dll so that we can use this in all modules, instead of having 2 or 3 different versions of this file.

Shims and verifier dll's cannot import from kernel32!

What is the reason to add the kernel32 dependency back?

**Review by learn-more — CHANGES_REQUESTED** [review]
Please verify that both shims and verifier dll's still work correctly, and do NOT link against kernel32.

File: dll/win32/mswsock/nsplookup.c [hunk 34]
```diff
@@ -14,8 +14,8 @@
 
 #include "mswhelper.h"
 
-#define NDEBUG
-#include <debug.h>
+#include <assert.h>
+#define ASSERT assert
```
Comment (ThFabba) [inline dll/win32/mswsock/nsplookup.c:18]: Not a fan of this. `msvcrt`'s assert results in annoying popups instead of just breaking into the debugger. Using it instead of `RtlAssert` may result in testbot timeouts without a backtrace, rather than an immediate debugger break that our test infrastructure knows how to handle.

File: sdk/include/reactos/debug.h (hunk 33 above)
Comment (ThFabba) [reply sdk/include/reactos/debug.h:18]: Yeah I'm not a huge fan of the idea, either. `ntdll` is always safe to import from, `kernel32` is not. So you're adding extra complexity to the code just to avoid ntdll imports. Because it's "cleaner", by some definition thereof?
Dependencies on ntdll don't hurt because it's always loaded in every process anyway. This fact isn't officially documented for Win32 applications, but (a) it's not something Microsoft can change at this point without breaking a ton of apps, (b) we're writing an OS anyway and a ton of our code relies on undocumented OS internals.

An argument can be made that applications should use kernel32 instead of ntdll, because they make sense on their own, even if you take them away from ROS. But for system DLLs (which seems to be what most of the affected modules here are) I see more potential problems than benefits.

Comment (katahiromz) [reply sdk/include/reactos/debug.h:18]: `IsDebuggerPresent` and `OutputDebugString` are of `kernel32`.

**Comment by katahiromz** [comment]
> do NOT link against kernel32.

Impossible. Closed.
